dotnet/src/Connectors/Connectors.Memory.AzureCosmosDBMongoDB/AzureCosmosDBMongoDBVectorStoreRecordCollection.cs

```diff
@@ -435,5 +514,115 @@ private static Dictionary<string, string> GetStoragePropertyNames(
         return storagePropertyNames;
     }
 
+    /// <summary>
+    /// Get vector property to use for a search by using the storage name for the field name from options
+    /// if available, and falling back to the first vector property in <typeparamref name="TRecord"/> if not.
+    /// </summary>
+    /// <param name="optionsVectorFieldName">The vector field name provided via options.</param>
+    /// <exception cref="InvalidOperationException">Thrown if the provided field name is not a valid field name.</exception>
+    private VectorStoreRecordVectorProperty? GetVectorPropertyForSearch(string? optionsVectorFieldName)
+    {
+        // If vector property name is provided in options, try to find it in schema or throw an exception.
+        if (!string.IsNullOrWhiteSpace(optionsVectorFieldName))
+        {
+            var vectorDataModelPropertyNameExists = this._dataModelPropertyNames
+                .TryGetValue(optionsVectorFieldName!, out var vectorDataModelPropertyName);
+
+            // Check vector properties by data model property name and storage property name.
+            var vectorProperty = this._vectorProperties
+                .FirstOrDefault(l => l.DataModelPropertyName.Equals(optionsVectorFieldName, StringComparison.Ordinal) ||
+                    (vectorDataModelPropertyNameExists && l.DataModelPropertyName.Equals(vectorDataModelPropertyName, StringComparison.Ordinal)));
+
+            if (vectorProperty is not null)
+            {
+                return vectorProperty;
+            }
+
+            throw new InvalidOperationException($"The collection does not have a vector property named '{optionsVectorFieldName}'.");
+        }
+
+        // If vector property is not provided in options, return first vector property from schema.
+        return this._firstVectorProperty;
+    }
+
+    /// <summary>Returns search part of the search query for <see cref="IndexKind.Hnsw"/> index kind.</summary>
+    private static BsonDocument GetSearchQueryForHnswIndex<TVector>(
+        TVector vector,
+        string vectorPropertyName,
+        int limit,
+        int efSearch)
+    {
+        return new BsonDocument
+        {
+            { "$search",
+                new BsonDocument
+                {
+                    { "cosmosSearch",
+                        new BsonDocument
+                        {
+                            { "vector", BsonArray.Create(vector) },
+                            { "path", vectorPropertyName },
+                            { "k", limit },
+                            { "efSearch", efSearch }
+                        }
+                    }
+                }
+            }
+        };
+    }
+
+    /// <summary>Returns search part of the search query for <see cref="IndexKind.IvfFlat"/> index kind.</summary>
+    private static BsonDocument GetSearchQueryForIvfIndex<TVector>(
```

Comment:
This file is pretty big. Consider splitting some of these helpers into a separate internal static class.